docs/proposals/20-ha-control-planes.md

```diff
@@ -0,0 +1,1259 @@
+Active-Active means that each component replica serves incoming requests (primarily intended for load balancing) whereas Active-Passive means that only one replica is active while others remain on stand-by.
+
+### Recommended number of nodes and zones
+
+It is recommended that for high-availability setup an odd number of nodes (node tolerance) or zones (zone tolerance) must be used. This also follows the [recommendations](https://etcd.io/docs/v3.4/faq/#why-an-odd-number-of-cluster-members) on the etcd cluster size. The recommendations for number of zones will be largely influenced by quorum-based etcd cluster setup recommendations as other shoot control plane components are either stateless or non-quorum-based stateful components.
+
+**Let's take the following example to explain this recommendation further:**
+* Seed clusters' worker nodes are spread across **two zones**
+* Gardener would distribute a **three member** etcd cluster - AZ-1: 2 replicas, AZ-2: 1 replica
+* If AZ-1 goes down then, quorum is lost and the only remaining etcd member enters into a read-only state.
+* If AZ-2 goes down then:
+  * If the leader is in AZ-2, then it will force a re-election and the quorum will be restored with 2 etcd members in AZ-1.
+  * If the leader is not in AZ-2, then etcd cluster will still be operational without any downtime as the quorum is not lost.
+
+**Result:**
+* There seems to be no clear benefit to spreading an etcd cluster across 2 zones as there is an additional cost of cross-zonal traffic that will be incurred due to communication amongst the etcd members and also due to API server communication with an etcd member across zones.
+* There is no significant gain in availability as compared to an etcd cluster provisioned within a single zone. Therefore it is a recommendation that for regions having 2 availability zones, etcd cluster should only be spread across nodes in a single AZ.
+
+### Recommended number of replicas
+
+The minimum number of replicas required to achieve HA depends on the [topology](#topologies) and the requirement of each component that run in an `active-active` mode.
+
+**Active-Active**
+* If application needs a quorum to operate (e.g. `etcd`), at least **three replicas** are required ([ref](https://etcd.io/docs/v3.3/faq/#why-an-odd-number-of-cluster-members)).
+* Non-quorum based components are also supposed to run with a minimum count of **three replicas** to survive node/zone outages and support load balancing.
+
+**Active-Passive**
+* Components running in  a `active-passive` mode are expected to have at least **two** replicas, so that there is always one replica on stand-by.
+
+## Gardener Shoot API
+
+### Proposed changes
+
+The following changes to the shoot API are suggested to enable the HA feature for a shoot cluster:
+
+```yaml
+kind: Shoot
+apiVersion: core.gardener.cloud/v1beta1
+spec:
+  controlPlane:
+    highAvailability:
```

Comment:
> For example that the existing alpha.control-plane.shoot.gardener.cloud/high-availability annotation will be dropped in favour of the new fields in the Shoot spec. Will this feature be introduced behind the existing HAControlPlanes feature gate?

Yes, we'll add this?

> https://github.com/gardener/gardener/pull/5741 was also added without any docs for end users or operators. I find it hard for myself figuring out with parts of this GEP are already covered by https://github.com/gardener/gardener/pull/5741 and which parts are about to be added/adapted. If you can add few sentences explaining what was already tackled with https://github.com/gardener/gardener/pull/5741 would be helpful for better understanding what is the current state and what we aim to achieve.

Hm, I'm trying to understand how this can help for the GEP. This GEP describes the overall target, no matter what is already implemented with the alpha feature gate `HAControlPlanes`.

Reply:
> Hm, I'm trying to understand how this can help for the GEP. This GEP describes the overall target, no matter what is already implemented with the alpha feature gate `HAControlPlanes`.

You could at least mention that https://github.com/gardener/gardener/pull/5741 adds the feature in alpha stage and this GEP builds on top of it - just to give context to the reader that something is added in the Gardener codebase. At least I was looking for such information/sentence and I would find it useful.

Reply:
Ok this can be added. There will be changes done to what has been already added though.

Reply:
There is a reference to the PR at  "Shoot Control Plane Components => Etcd". Is that good enough?